### PR TITLE
Create a C++ version of aprox19 + NSE

### PR DESCRIPTION
**Pull request by zingale**
This mirrors the Fortran aprox19 + NSE network

### CODE REVIEW

**Comment by zingale** [comment]
blocked by #327

File: integration/VODE/vode_type.H [hunk 1]
```diff
@@ -245,6 +249,26 @@ void update_thermodynamics (burn_t& state, dvode_t& vode_state)
 
     eos_t eos_state;
 
+#if NSE_THERMO
```
Comment (maxpkatz) [inline integration/VODE/vode_type.H:252]: ifdef

Comment (zingale) [reply integration/VODE/vode_type.H]: fixed

File: networks/aprox19/actual_network.H [hunk 2]
```diff
@@ -33,9 +33,9 @@ namespace C
 }
 
 #ifndef NSE
-const std::string network_name = "aprox19";
```
Comment (maxpkatz) [inline networks/aprox19/actual_network.H:36]: This looks like it was correct before

File: networks/aprox19/nse.H [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+#include <AMReX_Print.H>
+#include <AMReX_Algorithm.H>
+#include <AMReX_Array.H>
+#include <AMReX_REAL.H>
+
+#include "extern_parameters.H"
+
+#include "burn_type.H"
+
+#ifndef NSE_H
```
Comment (maxpkatz) [inline networks/aprox19/nse.H:15]: I'd prefer if the header guard is at the top of the file

Comment (zingale) [reply networks/aprox19/nse.H]: fixed

File: networks/aprox19/actual_network.H (hunk 2 above)
Comment (zingale) [reply networks/aprox19/actual_network.H]: yeah, that was some merge funniness.  Fixed

File: networks/aprox19/nse.H [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+  if (state.rho > rho_nse && state.T > T_nse) {
+
+    // Ma checks on Fe-group (for our composition, this means Cr48, Fe52, Fe54, Ni56)
+    // and C-group (for us, that is C12, N14)
+    // and He-group (for us, that is H1, He3, He4)
+
+    Real Fe_group = state.xn[Cr48-1] + state.xn[Fe52-1] +
+                    state.xn[Fe54-1] + state.xn[Ni56-1];
+    Real C_group = state.xn[C12-1] + state.xn[N14-1];
+    Real He_group = state.xn[H1-1] + state.xn[He3] + state.xn[He4];
```
Comment (maxpkatz) [inline networks/aprox19/nse.H:68]: Do these latter two accesses need -1 offsets too?

File: networks/aprox19_nse/Make.package [hunk 5]
```diff
@@ -0,0 +1,28 @@
+F90EXE_sources += actual_network.F90
```
Comment (maxpkatz) [inline networks/aprox19_nse/Make.package:1]: Looks like this file was a bad merge result too

File: unit_test/test_react/variables.F90 [hunk 6]
```diff
@@ -1,144 +0,0 @@
-!
```
Comment (maxpkatz) [inline unit_test/test_react/variables.F90:1]: Just confirming whether this deletion was intentional

Comment (zingale) [reply unit_test/test_react/variables.F90:1]: yes, it is not variables_F.F90

File: networks/aprox19_nse/Make.package (hunk 5 above)
Comment (zingale) [reply networks/aprox19_nse/Make.package]: fixed

File: networks/aprox19/nse.H (hunk 4 above)
Comment (zingale) [reply networks/aprox19/nse.H]: yep, good catch. fixed

**Comment by maxpkatz** [comment]
How does the Fortran versus C++ diff look in the unit tests?

**Comment by zingale** [comment]
I'll rerun the unit tests

**Comment by zingale** [comment]
the standalone test `aprox19/test-cpp/` just tests reading the table and the call to `nse_interp` in C++ and Fortran.  The results agree between the two interfaces:
```
Fortran: 
 temp:       1000000000.0000000     
 rho:        1000000000.0000000     
 ye:        0.46000000000000002     
 abar:       56.750299999999996     
 be/a:       8.7865299999999991     
 dyedt:     -2.4380900000002322E-006
 X:          0.0000000000000000        9.2078000000000612E-024   2.1277100000000334E-009   9.3008100000013293E-020   0.0000000000000000        2.6708400000009762E-019   7.0167400000050730E-023   5.8446400000017790E-018   2.9089200000003567E-012   1.0282800000000446E-010   2.2830700000002421E-010   1.4506700000004454E-011   5.5088400000005056E-009   6.7701299999999411E-004   5.7407199999999742E-002  0.94191600000000020        1.7993800055198185E-013   6.2451599999999448E-014   3.2610800000008766E-013
C++: 
temp:    1000000000
rho:     1000000000
ye:      0.46
abar:    56.7503
be/a:    8.78653
dyedt:   -2.43809e-06
X:       0 9.2078e-24 2.12771e-09 9.30081e-20 0 2.67084e-19 7.01674e-23 5.84464e-18 2.90892e-12 1.02828e-10 2.28307e-10 1.45067e-11 5.50884e-09 0.000677013 0.0574072 0.941916 1.799380006e-13 6.24516e-14 3.26108e-13 
```

**Comment by zingale** [comment]
Here are the `test_react` differences between C++ and Fortran, using `inputs_aprox19_nse` (compiled as: `make USE_CXX_REACTIONS=TRUE USE_NSE=TRUE NETWORK_DIR=aprox19 DEBUG=TRUE -j 20`)

```
            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 density                                          0                         0
 temperature                                      0                         0
 Xnew_hydrogen-1                      0.01230176153               0.225201379
 Xnew_helium-3                      5.162282384e-05            0.001253697158
 Xnew_helium-4                        0.03087627419             0.03817490673
 Xnew_carbon-12                      0.002215980598             0.05381667166
 Xnew_nitrogen-14                      0.0310207869              0.3806478917
 Xnew_oxygen-16                      0.002540173519             0.02682072139
 Xnew_neon-20                         0.02056316381              0.3670346306
 Xnew_magnesium-24                    0.01235936701              0.2645755572
 Xnew_silicon-28                      0.07266063448              0.2607432271
 Xnew_sulfur-32                       0.02860132722              0.2616011763
 Xnew_argon-36                       0.002373559882             0.03903208578
 Xnew_calcium-40                     0.007506001386             0.06559745189
 Xnew_titanium-44                     0.01571243989              0.3814233761
 Xnew_chromium-48                      0.0158404704              0.3769106905
 Xnew_iron-52                         0.01589957448              0.1892239761
 Xnew_iron-54                         0.09108238482               0.109932652
 Xnew_nickel-56                       0.01589212743              0.0159586931
 Xnew_neutron                        0.002732090806             0.03695495033
 Xnew_proton                         0.008994158694              0.2167572623
 Xold_hydrogen-1                                  0                         0
 Xold_helium-3                                    0                         0
 Xold_helium-4                                    0                         0
 Xold_carbon-12                                   0                         0
 Xold_nitrogen-14                                 0                         0
 Xold_oxygen-16                                   0                         0
 Xold_neon-20                                     0                         0
 Xold_magnesium-24                                0                         0
 Xold_silicon-28                                  0                         0
 Xold_sulfur-32                                   0                         0
 Xold_argon-36                                    0                         0
 Xold_calcium-40                                  0                         0
 Xold_titanium-44                                 0                         0
 Xold_chromium-48                                 0                         0
 Xold_iron-52                                     0                         0
 Xold_iron-54                                     0                         0
 Xold_nickel-56                                   0                         0
 Xold_neutron                                     0                         0
 Xold_proton                                      0                         0
 Xnew_Ye                              0.01096002174             0.02085243514
 Xnew_abar                             0.7739606986              0.0138299209
 Xnew_bea                              0.3032496804             0.03581688022
 Xold_Ye                                          0                         0
 Xold_abar                                        0                         0
 Xold_bea                                         0                         0
 wdot_hydrogen-1                        12301761.53              0.2987570665
 wdot_helium-3                          51622.82384            0.001253697523
 wdot_helium-4                          30876274.19             0.03859534432
 wdot_carbon-12                         2215980.598             0.05382447085
 wdot_nitrogen-14                        31020786.9              0.7533619695
 wdot_oxygen-16                         2540173.519             0.04745070293
 wdot_neon-20                           20563163.81              0.5012728514
 wdot_magnesium-24                      12359367.01              0.3230140413
 wdot_silicon-28                        72660634.48              0.3059512117
 wdot_sulfur-32                         28601327.22              0.3907426853
 wdot_argon-36                          2373559.882             0.05894108058
 wdot_calcium-40                        7506001.386               0.080349593
 wdot_titanium-44                       15712439.89              0.3817250399
 wdot_chromium-48                        15840470.4              0.3848097684
 wdot_iron-52                           15899574.48              0.2034681821
 wdot_iron-54                           91082384.82              0.2386613904
 wdot_nickel-56                         15892127.43             0.01605352111
 wdot_neutron                           2732090.806             0.08341285003
 wdot_proton                            8994158.694              0.2184341335
 rho_Hnuc                           1.421592943e+35              0.6725507681
```

**Comment by zingale** [comment]
the errors are large, but they are not coming from NSE regions.  If I if-def out the `valid_update` logic in `vode_dvstep.H`, then I get the following more reasonable diffs:
```
            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 density                                          0                         0
 temperature                                      0                         0
 Xnew_hydrogen-1                     2.70716278e-08           4.955849531e-07
 Xnew_helium-3                      1.790737394e-08           4.348933698e-07
 Xnew_helium-4                      6.726028846e-07           8.315949076e-07
 Xnew_carbon-12                     2.279048185e-08           5.534831307e-07
 Xnew_nitrogen-14                   8.137093262e-08           9.984812456e-07
 Xnew_oxygen-16                      3.33636543e-07           3.522740748e-06
 Xnew_neon-20                       3.162286243e-07           5.644406539e-06
 Xnew_magnesium-24                  2.330173915e-07           4.988176672e-06
 Xnew_silicon-28                    4.977711399e-07           1.786255436e-06
 Xnew_sulfur-32                     4.242111732e-07           3.880034694e-06
 Xnew_argon-36                      1.871809108e-07            3.07810282e-06
 Xnew_calcium-40                    3.692267531e-07           3.226795857e-06
 Xnew_titanium-44                   3.357471051e-07           8.150344264e-06
 Xnew_chromium-48                   4.308465422e-07           1.025163165e-05
 Xnew_iron-52                       1.532298943e-08           1.823619235e-07
 Xnew_iron-54                       4.275026861e-06           5.159779701e-06
 Xnew_nickel-56                     4.031491227e-06           4.048377507e-06
 Xnew_neutron                       1.211312739e-07           1.638452207e-06
 Xnew_proton                        1.350240833e-07           3.254050949e-06
 Xold_hydrogen-1                                  0                         0
 Xold_helium-3                                    0                         0
 Xold_helium-4                                    0                         0
 Xold_carbon-12                                   0                         0
 Xold_nitrogen-14                                 0                         0
 Xold_oxygen-16                                   0                         0
 Xold_neon-20                                     0                         0
 Xold_magnesium-24                                0                         0
 Xold_silicon-28                                  0                         0
 Xold_sulfur-32                                   0                         0
 Xold_argon-36                                    0                         0
 Xold_calcium-40                                  0                         0
 Xold_titanium-44                                 0                         0
 Xold_chromium-48                                 0                         0
 Xold_iron-52                                     0                         0
 Xold_iron-54                                     0                         0
 Xold_nickel-56                                   0                         0
 Xold_neutron                                     0                         0
 Xold_proton                                      0                         0
 Xnew_Ye                            5.993306884e-08           1.140280978e-07
 Xnew_abar                          3.341565933e-05           5.971051582e-07
 Xnew_bea                           1.492855577e-06           1.763214698e-07
 Xold_Ye                                          0                         0
 Xold_abar                                        0                         0
 Xold_bea                                         0                         0
 wdot_hydrogen-1                        27.07162778           6.574538191e-07
 wdot_helium-3                          17.90737394           4.348934965e-07
 wdot_helium-4                          672.6028846           8.407536403e-07
 wdot_carbon-12                         22.79048185           5.535633423e-07
 wdot_nitrogen-14                       81.37093262           1.976151225e-06
 wdot_oxygen-16                          333.636543           6.232364982e-06
 wdot_neon-20                           316.2286243           7.708776023e-06
 wdot_magnesium-24                      233.0173915           6.089946941e-06
 wdot_silicon-28                        497.7711399           2.095958623e-06
 wdot_sulfur-32                         424.2111732           5.795444795e-06
 wdot_argon-36                          187.1809108           4.648142745e-06
 wdot_calcium-40                        369.2267531           3.952466542e-06
 wdot_titanium-44                       335.7471051           8.156790287e-06
 wdot_chromium-48                       430.8465422           1.046647946e-05
 wdot_iron-52                           15.32298943           1.960895751e-07
 wdot_iron-54                           4275.026861           1.120176922e-05
 wdot_nickel-56                         4031.491227           4.072433337e-06
 wdot_neutron                           121.1312739           3.698231685e-06
 wdot_proton                            135.0240833           3.279224843e-06
 rho_Hnuc                           2.666887441e+29           1.261695343e-06
```

File: integration/VODE/vode_type.H [hunk 7]
```diff
@@ -171,10 +171,158 @@ void print_state(dvode_t& dvode_state)
...
+    // VODE state (which has the current xn, T, and e).
+
+    eos_t eos_state;
+
+#ifdef NSE_THERMO
+    // we are handling the thermodynamics via the aux quantities, which
+    // are stored in the burn_t here, so we need to update those based on
+    // the current state.
+
+    state.abar = 0.0;
```
Comment (harpolea) [inline integration/VODE/vode_type.H:262]:
```suggestion
    state.abar = 0.0_rt;
```

File: integration/VODE/vode_type.H [hunk 8]
```diff
@@ -171,10 +171,158 @@ void print_state(dvode_t& dvode_state)
...
+    state.y_e = 0.0;
```
Comment (harpolea) [inline integration/VODE/vode_type.H:263]:
```suggestion
    state.y_e = 0.0_rt;
```

File: networks/aprox19/nse.H [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+/// aux variables to be consistent with the new X's resulting from the burn.
+///
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void set_nse_aux_from_X(burn_t& state) {
+
+  using namespace AuxZero;
+
+  state.aux[iye] = 0.0;
+  state.aux[iabar] = 0.0;
+  state.aux[ibea] = 0.0;
```
Comment (harpolea) [inline networks/aprox19/nse.H:90]:
```suggestion
  state.aux[iye] = 0.0_rt;
  state.aux[iabar] = 0.0_rt;
  state.aux[ibea] = 0.0_rt;
```

**Review by harpolea — CHANGES_REQUESTED** [review]
This looks ok to me except for the header file with the merge conflict stuff in it

File: integration/VODE/vode_type.H [hunk 10]
```diff
@@ -171,10 +171,158 @@ void print_state(dvode_t& dvode_state)
 }
 #endif
 
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (zingale) [reply integration/VODE/vode_type.H]: I fixed this now.  Those changes moved to vode_type_strang.H

**Review by harpolea — APPROVED** [review]

**Pull request merged**